src/net_processing.cpp

```diff
@@ -344,8 +346,10 @@ struct CNodeState {
          */
         std::multimap<std::chrono::microseconds, uint256> m_tx_process_time;
 
-        //! Store all the transactions a peer has recently announced
-        std::set<uint256> m_tx_announced;
+        /* Store all the transactions a peer has recently announced,
```

Comment:
We discussed this in PR Review Club (https://bitcoincore.reviews/18238.html#l-115 and https://bitcoincore.reviews/18238.html#l-180), and there was some agreement that even for a small struct like this, the dependencies between the fields make it difficult to reason about and potentially bug-prone.

I think from a high level, in the future we might want a `TxDownloadState` class which contains:

- {txid, timestamp, state} objects where
    - state can be 'announced' or 'requested'
    - timestamp refers to 'when to request' if state is announced, and 'when requested' if state is requested
- a way to lookup the object by txid
- a way to iterate through the 'announced' objects sorted by timestamp
- a way to iterate through the 'requested' objects (doesn't necessarily need to be sorted by timestamp since there can only be MAX_PEER_TX_IN_FLIGHT = 100 of them)
- public functions to add, remove, refresh timestamps, etc

Reply:
I'm +1 on this, but maybe it'd be better to get wtxid relay #18044 in first?

Reply:
I've had a first attempt at this: https://github.com/jnewbery/bitcoin/tree/2020-03-tx-download-class

No tests or comments yet, and it could be tidied up, but do you think this is heading in the right direction?

Reply:
I was having a go at it too when I saw your comment, https://github.com/ajtowns/bitcoin/commits/202002-bump-notfound-wip -- separates out the current code into the class first, before changing it, which I think works better. Might be good to have some comments/style nits on the refactor?